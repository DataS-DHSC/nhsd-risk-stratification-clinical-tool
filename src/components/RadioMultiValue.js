import React from 'react';
import { Fieldset, Radios } from 'nhsuk-react-components';
import PropTypes from 'prop-types';

const RadioMultiValue = ({ field, groupName, prevData, setField }) => (
  <Fieldset>
    <Fieldset.Legend
      id={`${field.label.toLowerCase().replace(/ /g, '-')}-radios-label`}
    >
      {field.label}
    </Fieldset.Legend>

    <Radios
      id={groupName}
      name={groupName}
      hint={field.hint || ''}
      inline={field.inline}
    >
      {field.options.map((option) => (
        <Radios.Radio
          key={`${groupName}-${option.name}`}
          id={`${groupName}-${option.name}`}
          value={`${option.name}`}
          defaultChecked={
            prevData[groupName.toString()]
              ? prevData[groupName.toString()] === `${option.name}`
              : field.default === `${option.name}`
          }
          required
          onClick={() => setField(groupName, option.name)}
        >
          {option.label}
        </Radios.Radio>
      ))}
    </Radios>
  </Fieldset>
);
RadioMultiValue.propTypes = {
  field: PropTypes.shape().isRequired,
  prevData: PropTypes.shape(),
  groupName: PropTypes.string.isRequired,
  setField: PropTypes.func.isRequired,
};

RadioMultiValue.defaultProps = {
  prevData: {},
};

export default RadioMultiValue;
