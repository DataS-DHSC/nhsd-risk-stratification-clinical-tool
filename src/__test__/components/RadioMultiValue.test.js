import React from 'react';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import RadioMultiValue from '../../components/RadioMultiValue';

const setupDefaultProps = (defaultFieldValue) => {
  const props = {
    field: {
      label: 'Diabetes',
      type: 'multiradio',
      inline: false,
      default: defaultFieldValue,
      options: [
        { name: 'value1', label: 'Value1' },
        { name: 'value2', label: 'Value2' },
        { name: 'value3', label: 'Value3' },
      ],
    },
    groupName: 'diabetes',
    prevData: {},
    setField: jest.fn(),
  };

  return props;
};

describe('RadioMultiValue', () => {
  it('should take its value from prevData rather than the default', () => {
    const props = {
      ...setupDefaultProps('value2'),
      prevData: { diabetes: 'value1' },
    };

    const container = render(<RadioMultiValue {...props} />);
    expect(container.asFragment()).toMatchSnapshot();
  });

  it('should take the default value when the field is not in prevData', () => {
    const props = { ...setupDefaultProps('value2') };
    const container = render(<RadioMultiValue {...props} />);
    expect(container.asFragment()).toMatchSnapshot();
  });

  it('should should call setField on value change', () => {
    const setField = jest.fn();

    const props = {
      ...setupDefaultProps('value1'),
      setField,
    };

    const { container } = render(<RadioMultiValue {...props} />);

    userEvent.click(container.querySelector('#diabetes-value2'));

    expect(setField).toHaveBeenCalledWith('diabetes', 'value2');
  });
});
