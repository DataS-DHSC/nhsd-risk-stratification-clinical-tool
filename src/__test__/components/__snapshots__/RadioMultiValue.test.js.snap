// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioMultiValue should take its value from prevData rather than the default 1`] = `
<DocumentFragment>
  <div
    class="nhsuk-form-group"
  >
    <fieldset
      class="nhsuk-fieldset"
    >
      <legend
        class="nhsuk-fieldset__legend"
        id="diabetes-radios-label"
      >
        Diabetes
      </legend>
      <div
        class="nhsuk-form-group"
      >
        <div
          class="nhsuk-radios"
          id="diabetes"
          role="radiogroup"
        >
          <div
            class="nhsuk-radios__item"
          >
            <input
              aria-labelledby="diabetes-value1--label"
              checked=""
              class="nhsuk-radios__input"
              id="diabetes-value1"
              name="diabetes"
              required=""
              type="radio"
              value="value1"
            />
            <label
              class="nhsuk-label nhsuk-radios__label"
              for="diabetes-value1"
              id="diabetes-value1--label"
            >
              Value1
            </label>
          </div>
          <div
            class="nhsuk-radios__item"
          >
            <input
              aria-labelledby="diabetes-value2--label"
              class="nhsuk-radios__input"
              id="diabetes-value2"
              name="diabetes"
              required=""
              type="radio"
              value="value2"
            />
            <label
              class="nhsuk-label nhsuk-radios__label"
              for="diabetes-value2"
              id="diabetes-value2--label"
            >
              Value2
            </label>
          </div>
          <div
            class="nhsuk-radios__item"
          >
            <input
              aria-labelledby="diabetes-value3--label"
              class="nhsuk-radios__input"
              id="diabetes-value3"
              name="diabetes"
              required=""
              type="radio"
              value="value3"
            />
            <label
              class="nhsuk-label nhsuk-radios__label"
              for="diabetes-value3"
              id="diabetes-value3--label"
            >
              Value3
            </label>
          </div>
        </div>
      </div>
    </fieldset>
  </div>
</DocumentFragment>
`;

exports[`RadioMultiValue should take the default value when the field is not in prevData 1`] = `
<DocumentFragment>
  <div
    class="nhsuk-form-group"
  >
    <fieldset
      class="nhsuk-fieldset"
    >
      <legend
        class="nhsuk-fieldset__legend"
        id="diabetes-radios-label"
      >
        Diabetes
      </legend>
      <div
        class="nhsuk-form-group"
      >
        <div
          class="nhsuk-radios"
          id="diabetes"
          role="radiogroup"
        >
          <div
            class="nhsuk-radios__item"
          >
            <input
              aria-labelledby="diabetes-value1--label"
              class="nhsuk-radios__input"
              id="diabetes-value1"
              name="diabetes"
              required=""
              type="radio"
              value="value1"
            />
            <label
              class="nhsuk-label nhsuk-radios__label"
              for="diabetes-value1"
              id="diabetes-value1--label"
            >
              Value1
            </label>
          </div>
          <div
            class="nhsuk-radios__item"
          >
            <input
              aria-labelledby="diabetes-value2--label"
              checked=""
              class="nhsuk-radios__input"
              id="diabetes-value2"
              name="diabetes"
              required=""
              type="radio"
              value="value2"
            />
            <label
              class="nhsuk-label nhsuk-radios__label"
              for="diabetes-value2"
              id="diabetes-value2--label"
            >
              Value2
            </label>
          </div>
          <div
            class="nhsuk-radios__item"
          >
            <input
              aria-labelledby="diabetes-value3--label"
              class="nhsuk-radios__input"
              id="diabetes-value3"
              name="diabetes"
              required=""
              type="radio"
              value="value3"
            />
            <label
              class="nhsuk-label nhsuk-radios__label"
              for="diabetes-value3"
              id="diabetes-value3--label"
            >
              Value3
            </label>
          </div>
        </div>
      </div>
    </fieldset>
  </div>
</DocumentFragment>
`;
